pycolleff/pycolleff/longitudinal_equilibrium.py

```diff
@@ -1335,5 +1718,131 @@ def _feedback_err(x, *args):
+
+    @staticmethod
+    def _forest_ruth_integrator(z0, p0, ds, alpha, zgrid, vtotal):
+        """4th-order symplectic integrator using the Forest-Ruth method."""
+        z, p = z0, p0
+        positions = [z0]
+        momentums = [p0]
+        angle = 0
+        theta = 1 / (2 - 2 ** (1 / 3))
+
+        elapsed = 0
+        while True:
+            # step 1
+            z += alpha * p * theta * ds / 2
+            p += _np.interp(z, zgrid, vtotal) * theta * ds
+
+            # step 2
+            z += alpha * p * (1 - theta) * ds / 2
+            p += _np.interp(z, zgrid, vtotal) * (1 - 2 * theta) * ds
+
+            # step 3
+            z += alpha * p * (1 - theta) * ds / 2
+            p += _np.interp(z, zgrid, vtotal) * theta * ds
+
+            # step 4
+            z += alpha * p * theta * ds / 2
+
+            dangle = LongitudinalEquilibrium._calc_dangle(
+                positions[-1], momentums[-1], z, p
+            )
+
+            angle += dangle
+            if angle > _2PI:
+                break
+
+            positions.append(z)
+            momentums.append(p)
+            elapsed += 1
+            if elapsed > 1e6:
+                raise Exception("More than 1e6 steps in integrator.")
+        return positions, momentums
+
+    @staticmethod
+    def _position_extended_forest_ruth_integrator(
+        z0, p0, ds, alpha, zgrid, vtotal
+    ):
+        """Position Extended Forest-Ruth Like method."""
+        z, p = z0, p0
+        positions = [z0]
+        momentums = [p0]
+        angle = 0
+        xi = +0.1786178958448091
+        lamb = -0.212341831062605
+        chi = -0.0662645826698184
+
+        elapsed = 0
+        while True:
+            # step 1
+            z += alpha * p * xi * ds
+            p += _np.interp(z, zgrid, vtotal) * (1 - 2 * lamb) * ds / 2
+
+            # step 2
+            z += alpha * p * chi * ds
+            p += _np.interp(z, zgrid, vtotal) * lamb * ds
+
+            # step 3
+            z += alpha * p * (1 - 2 * (chi + xi)) * ds
+            p += _np.interp(z, zgrid, vtotal) * lamb * ds
+
+            # step 4
+            z += alpha * p * chi * ds / 2
+            p += _np.interp(z, zgrid, vtotal) * (1 - 2 * lamb) * ds / 2
+
+            # step 5
+            z += alpha * p * xi * ds
+
+            dangle = LongitudinalEquilibrium._calc_dangle(
+                positions[-1], momentums[-1], z, p
+            )
+
+            angle += dangle
+            if angle > _2PI:
+                break
+
+            positions.append(z)
+            momentums.append(p)
+
+            elapsed += 1
+            if elapsed > 1e6:
+                raise Exception("More than 1e6 steps in integrator.")
+        return positions, momentums
+
+    @staticmethod
+    def _calc_dangle(z0, p0, z, p):
+        vec0 = _np.array([z0, p0])
+        vec0 /= _np.linalg.norm(vec0)
+        vec1 = _np.array([z, p])
+        vec1 /= _np.linalg.norm(vec1)
+        acos = _np.abs(_np.dot(vec0, vec1))
```

Comment:
in this case, wouldn't be faster to just do:
```suggestion
        acos = z0*z + p0*p
        acos /= np.sqrt((z0 * z0 + p0 * p0) * (z*z + p*p))
```